File: demos/demo_testbench.cpp
#include <odri_control_interface/calibration.hpp>
#include <odri_control_interface/robot.hpp>

#include <odri_control_interface/utils.hpp>

using namespace odri_control_interface;

#include <iostream>
#include <stdexcept>

typedef Eigen::Matrix<double, 2, 1> Vector2d;

int main()
{
    nice(-20);  // Give the process a high priority.

    // Define the robot from a yaml file.
    auto robot = RobotFromYamlFile(CONFIG_TESTBENCH_YAML);

    // Store initial position data.
    Vector2d des_pos(3.1415 * 0.5, - 3.1415 * 0.5);

    // Initialize the communication, session, joints, wait for motors to be ready
    // and run the joint calibration.
    robot->Initialize(des_pos);

    // Initialize simple pd controller.
    Vector2d torques;
    double kp = 0.125;
    double kd = 0.0025;
    int c = 0;
    while (!robot->IsTimeout())
    {
        robot->ParseSensorData();

        // Run the main controller.
        auto pos = robot->joints->GetPositions();
        auto vel = robot->joints->GetVelocities();

        // Compute PD control on the zero position.
        for (int i = 0; i < 2; i++)
        {
            torques[i] = kp * (des_pos[i] - pos[i]) - kd * vel[i];
        }
        robot->joints->SetTorques(torques);

        // Checks if the robot is in error state (that is, if any component
        // returns an error). If there is an error, the commands to send
        // are changed to send the safety control.
        robot->SendCommandAndWaitEndOfCycle(0.001);

        c++;
        if (c % 1000 == 0)
        {
            std::cout << "Joints: ";
            robot->joints->PrintVector(robot->joints->GetPositions());
            std::cout << std::endl;
        }
    }

    if (robot->IsTimeout()) {
        printf("Timeout detected\n");
    } else if (robot->HasError()) {
        printf("Error detected\n");
    }

    return 0;
}

File: demos/demo_testbench.py
import numpy as np

np.set_printoptions(suppress=True, precision=2)

import libodri_control_interface_pywrap as oci

robot = oci.robot_from_yaml_file("config_testbench.yaml")

des_pos = np.array([np.pi / 2, -np.pi / 2])

robot.initialize(des_pos)

kp, kd = 0.125, 0.0025
c = 0
while not robot.is_timeout:
    robot.parse_sensor_data()

    positions = robot.joints.positions
    velocities = robot.joints.velocities

    # Compute the PD control on the zero position.
    torques = kp * (des_pos - positions) - kd * velocities

    robot.joints.set_torques(torques)
    robot.send_command_and_wait_end_of_cycle(0.001)

    c += 1

    if c % 1000 == 0:
        print("joint pos:   ", positions)
        print("joint vel:   ", velocities)
        print("torques:     ", torques)
        robot.robot_interface.PrintStats()

print("timeout detected")
